### ISSUE TITLE
cast() doesn't affect uses in different scopes

### ISSUE THREAD

**Issue by funketh** [python/mypy#7003]
```python
n: Optional[str] = 'World'
print('Hello ' + n)  # mypy complains
n = cast(str, n)
print('Hello ' + n)  # mypy doesn't complaint

def foo() -> None:
    print('Hello ' + n)  # mypy complains (even though n has been casted)

def bar() -> None:
    global n
    n = cast(str, n)
    print('Hello ' + n)  # mypy doesn't complain
```
I understand that mypy supposedly doesn't support type redefinitions but why then does `cast()` work in the first place?

This came up while refactoring code into its own method.
When calling a function which casts a nonlocal variable I expected the cast to take effect for any statements following the call to that funtion.

This is fine:
```python
class Hello:
    def __init__(self, name: Optional[str]):
        self.name = name
        self.ensure_string()
        self.name = cast(str, self.name)
        print('Hello ' + self.name)

    def ensure_string(self) -> None:
        self.name = self.name or ''
```

But here mypy complains:
```python
class Hello:
    def __init__(self, name: Optional[str]):
        self.name = name
        self.ensure_string()
        print('Hello ' + self.name)

    def ensure_string(self) -> None:
        self.name = self.name or ''
        self.name = cast(str, self.name)
```

**Comment by ilevkivskyi** [comment]
It looks like there are few misunderstandings. First, `cast()` normally applies to an expression not to a variable, so `n = cast(str, n)` is rather an anti-pattern, and should be probably replaced with `assert isinstance(n, str)`.

Second, mypy always does some type narrowing using local control flow. It also allows (most) redefinitions if you are using `--allow-redefinitions`.

Finally, a function/method is a black box for mypy, the only public information is its type, mypy can't reason about side effects of a method like making some other attribute non-optional.

There are couple related existing issues (making `--alow-redefinitions` work in conditional statements and support custom type guards) but I think there are no new action items here.
